src/uu/expr/src/syntax_tree.rs

```diff
@@ -173,6 +175,87 @@ impl StringOp {
+///
+/// GNU coreutils shows messages for invalid regular expressions
+/// differently from the oniguruma library used by the regex crate.
+/// This method attempts to do these checks manually in one pass
+/// through the regular expression.
+fn validate_regex(pattern: &str) -> ExprResult<()> {
+    let mut escaped_parens: u64 = 0;
+    let mut escaped_braces: u64 = 0;
+    let mut escaped = false;
+
+    let mut repeating_pattern_text = String::with_capacity(13);
+    let mut invalid_content_error = false;
+
+    for c in pattern.chars() {
+        match (escaped, c) {
+            (true, ')') => {
+                escaped_parens = escaped_parens
+                    .checked_sub(1)
+                    .ok_or(ExprError::UnmatchedClosingParenthesis)?;
+            }
+            (true, '(') => {
+                escaped_parens += 1;
+            }
+            (true, '}') => {
+                escaped_braces = escaped_braces
+                    .checked_sub(1)
+                    .ok_or(ExprError::UnmatchedClosingBrace)?;
+                let mut repetition = repeating_pattern_text[..repeating_pattern_text.len() - 1]
+                    .splitn(2, |x| x == ',');
+                match (repetition.next(), repetition.next()) {
+                    (None, None) => {
+                        // Empty repeating pattern
+                        invalid_content_error = true;
+                    }
+                    (Some(x), None) | (Some(x), Some("")) => {
+                        if x.parse::<i16>().is_err() {
+                            invalid_content_error = true;
+                        }
+                    }
+                    (None, Some(x)) | (Some(""), Some(x)) => {
+                        if x.parse::<i16>().is_err() {
+                            invalid_content_error = true;
+                        }
+                    }
+                    (Some(f), Some(l)) => {
+                        if let (Ok(f), Ok(l)) = (f.parse::<i16>(), l.parse::<i16>()) {
+                            invalid_content_error = invalid_content_error || f > l;
+                        } else {
+                            invalid_content_error = true;
+                        }
+                    }
+                }
+                repeating_pattern_text.clear();
+            }
+            (true, '{') => {
+                escaped_braces += 1;
+            }
+            _ => {
+                if escaped_braces > 0 && repeating_pattern_text.len() <= 13 {
+                    repeating_pattern_text.push(c);
+                }
+                if escaped_braces > 0 && !(c.is_ascii_digit() || c == '\\' || c == ',') {
+                    invalid_content_error = true;
+                }
+            }
+        }
+        escaped = !escaped && c == '\\';
+    }
+    match (
+        escaped_parens.is_zero(),
+        escaped_braces.is_zero(),
+        invalid_content_error,
+    ) {
+        (true, true, false) => Ok(()),
```

Comment:
Anti-pattern: This code halfway-parses the regex, then throws away most of the result, and then parses the regex from scratch in different code (I believe in line 145, `let re = Regex::with_options(`). Any slight mismatch between what `validate_regex` calls valid and what `Regex::with_options` calls valid is going to be a bug.

Is there any way to extract more error context from onig, instead of re-implementing a slightly different parser?

Reply:
Oh no, that is unfortunate :(

Yeah, this makes sens then. I can't come up with anything better.